src/darwin/Framework/CHIP/MTRLogging.mm

```diff
@@ -0,0 +1,174 @@
+
++ (instancetype)sharedInstance
+{
+    static OsLogMTRLogger * singleton = nil;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        // initialize the factory.
+        singleton = [[OsLogMTRLogger alloc] init];
+    });
+    return singleton;
+}
+
+- (instancetype)init
+{
+    return self;
+}
+
+- (BOOL)isLoggingEnabledAtLevel:(MTRLogLevel)level
+{
+    return true;
+}
+
+- (void)logAtLevel:(MTRLogLevel)level message:(NSString *)message
+{
+    switch (level) {
+    case MTRLogLevelError:
+        os_log_error(OS_LOG_DEFAULT, "%@", message);
+        break;
+    case MTRLogLevelInfo:
+        os_log_info(OS_LOG_DEFAULT, "%@", message);
+        break;
+    case MTRLogLevelDebug:
+        os_log_debug(OS_LOG_DEFAULT, "%@", message);
+        break;
+    default:
+        break;
+    }
+}
+
+@end
+
+// MARK: - MTRLogging
+
+@implementation MTRLoggingConfiguration
+
+static id<MTRLogger> singleton_logger;
+
+/**
+ * A log redirection function that will be used to capture logs from the Matter C++ SDK.
+ */
+static ENFORCE_FORMAT(3, 0) void cpp_log_redirect_callback(const char * module, uint8_t category, const char * msg, va_list args)
+{
+    char formatted_msg[CHIP_CONFIG_LOG_MESSAGE_MAX_SIZE];
+    chip::Logging::Platform::getDarwinLogMessageFormat(module, category, msg, args, CHIP_CONFIG_LOG_MESSAGE_MAX_SIZE, formatted_msg);
+
+    // Delegate the log to the MTRLogging system
+    switch (category) {
+    case chip::Logging::LogCategory::kLogCategory_Error:
+        [MTRLogging logAtLevel:MTRLogLevelError format:@"%s", formatted_msg];
+        break;
+
+    case chip::Logging::LogCategory::kLogCategory_Progress:
+        [MTRLogging logAtLevel:MTRLogLevelInfo format:@"%s", formatted_msg];
+        break;
+
+    case chip::Logging::LogCategory::kLogCategory_Detail:
+        [MTRLogging logAtLevel:MTRLogLevelDebug format:@"%s", formatted_msg];
+        break;
+
+    default:
+        [MTRLogging logAtLevel:MTRLogLevelInfo format:@"%s", formatted_msg];
+        break;
+    }
+}
+
++ (void)setLogger:(id<MTRLogger>)logger 
+{
+    singleton_logger = logger;
+    chip::Logging::SetLogRedirectCallback(cpp_log_redirect_callback);
+}
+
++ (id<MTRLogger>)getLogger
+{
+    if (singleton_logger == nil)
+    {
+        singleton_logger = [OsLogMTRLogger sharedInstance];
+    }
+    return singleton_logger;
+}
+
+@end
+
+@implementation MTRLogging
+
+// queue used to serialize all work performed by the MTRLogger
+static dispatch_queue_t chipWorkQueue;
+
++ (void)logAtLevel:(MTRLogLevel)level format:(NSString *)format, ... NS_FORMAT_FUNCTION(2, 3)
+{
+    va_list args;
+    va_start(args, format);
+    [MTRLogging logAtLevelV:level format:format args:args];
+    va_end(args);
+}
+
++ (void)logAtLevelV:(MTRLogLevel)level format:(NSString *)format args:(va_list)args NS_FORMAT_FUNCTION(2, 0)
+{
+    if (chipWorkQueue == nil)
+    {
+        chipWorkQueue = chip::DeviceLayer::PlatformMgrImpl().GetWorkQueue();
```

Comment:
No, the whole point is we do _not_ want to call out into consumer code on that work queue.

Reply:
Got it, created a new queue altogether.